ql/src/java/org/apache/hadoop/hive/ql/optimizer/calcite/rules/HiveOptimizeInlineArrayTableFunctionRule.java

```diff
@@ -0,0 +1,128 @@
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.hadoop.hive.ql.optimizer.calcite.rules;
+
+import org.apache.calcite.plan.RelOptCluster;
+import org.apache.calcite.plan.RelOptRule;
+import org.apache.calcite.plan.RelOptRuleCall;
+import org.apache.calcite.rel.RelNode;
+import org.apache.calcite.rel.type.RelDataType;
+import org.apache.calcite.rel.type.RelDataTypeField;
+import org.apache.calcite.rex.RexCall;
+import org.apache.calcite.rex.RexInputRef;
+import org.apache.calcite.rex.RexNode;
+import org.apache.calcite.tools.RelBuilderFactory;
+import org.apache.hadoop.hive.ql.optimizer.calcite.HiveRelFactories;
+import org.apache.hadoop.hive.ql.optimizer.calcite.reloperators.HiveTableFunctionScan;
+
+import com.google.common.base.Preconditions;
+import com.google.common.collect.Lists;
+
+import java.util.ArrayList;
+import java.util.List;
+
+/**
+ * This rule optimizes the inline udtf in a HiveTableFunctionScan when it
+ * has an array of structures. The RelNode for a HiveTableFunctionScan places
+ * the input references as the first elements in the return type followed by
+ * the udtf return value which represents the items in the generated table. Take the
+ * case where the base (input) table has col1, and the inline function is represented by:
+ * inline(array( struct1(col2, col3), struct2(col2, col3), struct3(col2, col3), etc...)),
+ * ...and the return value for the table scan node is (col1, col2, col3). In this case,
+ * the same col1 value is joined with the structures within the inline array for the
+ * col2 and col3 values.
+ *
+ * The optimization is to put the "col1" value within the inline array, resulting in
+ * in the new structure:
+ * inline(array(struct1(col1, col2, col3), struct2(col1, col2, col3), ...)
+ * By doing this, we avoid creating a lateral view join operator and a lateral view forward
+ * operator at runtime.
+ */
+public class HiveOptimizeInlineArrayTableFunctionRule extends RelOptRule {
+
+  public static final HiveOptimizeInlineArrayTableFunctionRule INSTANCE =
+          new HiveOptimizeInlineArrayTableFunctionRule(HiveRelFactories.HIVE_BUILDER);
+
+  public HiveOptimizeInlineArrayTableFunctionRule(RelBuilderFactory relBuilderFactory) {
+    super(operand(HiveTableFunctionScan.class, any()), relBuilderFactory, null);
+  }
+
+  @Override
+  public boolean matches(RelOptRuleCall call) {
+    final HiveTableFunctionScan tableFunctionScanRel = call.rel(0);
+
+    Preconditions.checkState(tableFunctionScanRel.getCall() instanceof RexCall);
```

Comment:
I think I like it this way because Hive should never allow anything other than a RexCall within the HiveTableFunctionScan.  I guess arguably I can get rid of this preconditions check because the very next statement would catch the exception anyway.  Not sure which is the preferred Hive way.

Reply:
I was thinking of something like this instead of precondition:
```
if (!(tableFunctionScanRel.getCall() instanceof RexCall)) {
  return false;
}
```
So we just silently don't apply this rule. Not sure whether we would end up with an invalid plan. If so then we can keep precondition otherwise it is too agressive.

Btw I saw assertions elsewhere in the code but those are not run in prod so the outcome still depends on the plan.